component.js

```diff
@@ -200,33 +206,35 @@ function factory (options) {
       debug = _shouldComponentUpdate.debug(pattern, logFn);
     }
   }
-
-  function createDefaultArguments (displayName, mixins, render) {
-
-    // (render)
-    if (typeof displayName === 'function') {
-      render      = displayName;
-      mixins      = [];
-      displayName = void 0;
+  
+  var NO_RENDER = "Component definition requires `render` function as its argument"
+
+  // [displayName], [mixins = []], render, [decorate]
+  function createDefaultArguments (/* args */) {
+    var args = Array.prototype.slice.call(arguments);
+    var currArg = 0, argCount = args.length;;
+    
+    var displayName, mixins = [], render, decorate;
+    
+    if (currArg == argCount) throw NO_RENDER;
+    if (typeof args[currArg] === 'string' || args[currArg] instanceof String) {
+      displayName = args[currArg++];
     }
-
-    // (mixins, render)
-    if (typeof displayName === 'object' && typeof mixins === 'function') {
-      render      = mixins;
-      mixins      = displayName;
-      displayName = void 0;
+    
+    if (currArg == argCount) throw NO_RENDER;
```

Comment:
I'm not eager to change it either, though I see no other way to let user pass a decorator without losing ability to pass `displayName` and `mixins`. Actually I don't like passing optional arguments at all (that's from a painful Haskell experience). Rewrite of this method is needed for this purpose, because otherwise there will be something like 8 checks, and neither of these are obvious.

I think the way it should work is like this.

![image](https://cloud.githubusercontent.com/assets/2983969/10455442/60730278-71cc-11e5-9b02-149173d8c1ef.png)

`$` is for "no more arguments" and `*` is for "any more arguments". Double circles are special states (initial or terminal). I've dropped the check for `*` because there're cases like `x.map(f.bind(...))` and some extra arguments may sneak into the code. Though I somehow failed to implement that FSM verbatim.